core/src/main/java/org/infinispan/interceptors/locking/PessimisticLockingInterceptor.java

```diff
@@ -1,126 +1,94 @@
- * in place.
- *
- * Implementation note: current implementation acquires locks remotely first and then locally. This is required
- * by the deadlock detection logic, but might not be optimal: acquiring locks locally first might help to fail fast the
- * in the case of keys being locked.
+ * Locking interceptor to be used by pessimistic transactions.
+ * <p>
+ * Design (from 9.1):
+ * <p>
+ * This interceptor is skipped when {@link Flag#SKIP_LOCKING} is used or when all affected keys are already locked.
+ * Read-only commands can acquire the locks when {@link Flag#FORCE_WRITE_LOCK}.
+ * <p>
+ * The new design has 2 concepts. The primary lock and the backup lock. The primary lock is acquired in the primary
+ * owner and it is exclusive; only one transaction can acquire it. The backup lock is acquired by the backup owner and
+ * it is shared; multiple transactions can acquire simultaneously. The backup locks are associated with the topology id
+ * and they are only used when the primary owner leaves or crashes.
+ * <p>
+ * This algorithm contacts all owners and acquires the primary lock in the primary owner and the backup lock in the
+ * backup owners. So, lock acquisition (single or multiple keys) happens in a single RPC to all the owners of the
+ * key(s). This is triggered in {@link TxDistributionInterceptor#visitLockControlCommand(TxInvocationContext,
+ * LockControlCommand)}.
+ * <p>
+ * The {@link Flag#CACHE_MODE_LOCAL} changes this behaviour; it makes the lock acquisition only local. So, only local
+ * the primary (if primary owner) or backup lock (if backup owner) is acquired. No RPC is create and if the node isn't
+ * an owner, no locks are acquired.
+ * <p>
+ * <p>
+ * In case of primary owner changes, there are 2 cases.
+ * <p>
+ * 1) a new node joins and it is the new primary owner. In this case, the old primary owner transfers the lock
+ * information to the new primary owner.
+ * <p>
+ * 2) the primary owner crashes or leaves. In this case, a backup owner will be promoted to the primary owner. All
+ * primary lock requests are hold until its previous backup locks are released. Note that a transaction only acquires
+ * the lock once. If a transaction requests a primary lock but has the backup lock acquired, it doesn't grant the
+ * primary lock if there are other backup locks from other transactions.
+ * <p>
+ * Know Issue: There is small deadlock probably is 2 or more transaction try to acquire the lock of a key and the
+ * primary owner dies before any of them receives a reply. In that case, none of the transactions (and future
+ * transactions) are able to acquire the primary lock until all backup lock are cleared. Eventually, the transactions
+ * will rollback and the backup lock will be released. One way to solve it is to acquire the backup locks after the
+ * primary lock. However, this will increase the performance penalty for all locks by generating 2 RPCs.
```

Comment:
IMO the originator should release the backup lock (only for this key, not for all the keys in the transaction) and retry if any it gets a `CacheNotFoundResponse` or `OutdatedTopologyException` from any of the targets.

Reply:
@danberindei You can't just release a lock and reacquire it - you'd be leaving a window to put in some value not observed by the modifying transaction. Unless we go with WSC in pessimistic caches, which would be my preference.

Reply:
@rvansa The lock release and re-acquisition would both happen before the `lock()` operation finished on the originator. If a transaction tries to acquire the lock after reading it, there's already a window for other transactions to modify the value.

Reply:
@danberindei OK, I thought you suggest to re-acquire the lock during commit.

Reply:
@rvansa No, that would really require WSC :)